### ISSUE TITLE
Growing performance penalty over time

### ISSUE THREAD

**Issue by mike-marcacci** [googleapis/cloud-debug-nodejs#814]
We have spent an extraordinary amount of time tracing progressive degradation of app performance (to the point of unresponsiveness) to this tool. The problem does not appear to be the result of a change in _this_ library, but a change in one of its direct or transitive _dependencies_, which made it quite painful to track down.

Beginning with the deployment of a service that included only small changes and updated dependencies, we began experiencing a significant leak of sorts. The latency of HTTP requests between our cloud load balancer and application would continue to grow, while becoming increasingly variable. At the same time, all metrics and samples collected from _within_ the application remained stable.

Through a lot of trial and error we discovered that we could only reproduce the issue in production, which made us suspect the various Google agents we have installed (trace, profile, and debug). We eliminated all of them, which fixed our issues. We first reintroduced trace without experiencing the leak; we then added profile without issue; we finally added debug back in and the behavior described here returned.

Here are a few additional deploys we did to test the effect of `allowExpressions`:

<img width="2007" alt="image" src="https://user-images.githubusercontent.com/690517/72780314-6cef9400-3bd3-11ea-8f4a-99dd757dbce1.png">

The latencies continue increasing about linearly until we restart the processes (or kill pods in our case).

Between "good" and "bad" versions of our application, the following dependencies of `@google-cloud/debug-agent` changed, so there's likely an interaction somewhere in one of these:

| name                      | last good version | current version |
|---------------------------|-------------------|-----------------|
| @google-cloud/common      | 2.0.4             | 2.2.6           |
| @google-cloud/debug-agent | 4.2.0             | 4.2.1           |
| @google-cloud/projectify  | 1.0.1             | 1.0.4           |
| @google-cloud/promisify   | 1.0.2             | 1.0.4           |
| base64-js                 | 1.3.0             | 1.3.1           |
| coffeescript              | 2.4.1             | 2.5.0           |
| end-of-stream             | 1.4.1             | 1.4.4           |
| gaxios                    | 2.0.1             | 2.2.2           |
| gcp-metadata              | 3.2.2             | 3.3.0           |
| google-auth-library       | 4.2.6             | 5.9.1           |
| google-p12-pem            | 2.0.1             | 2.0.4           |
| gtoken                    | 3.0.2             | 4.1.4           |
| https-proxy-agent         | 2.2.2             | 4.0.0           |
| jwa                       | 1.4.1             | 2.0.0           |
| jws                       | 3.2.2             | 4.0.0           |
| node-forge                | 0.8.5             | 0.9.1           |
| p-limit                   | 2.2.0             | 2.2.2           |
| readable-stream           | 2.3.6             | 2.3.7           |
| semver                    | 5.7.0             | 7.1.1           |
| stream-shift              | 1.0.0             | 1.0.1           |
| string_decoder            | 1.2.0             | 1.1.1           |
| teeny-request             | 4.0.0             | 6.0.0           |
| uuid                      | 3.3.2             | 3.4.0           |
| yallist                   | 3.0.3             | 3.1.1           |

We also observe a small memory leak, which may mean that this is the same as #811. However, the memory growth was very small and slow on node version 10.x, and was dwarfed by our concerns about latency. I will also note that when testing on node versions 12 and 13, we experienced a _significant_ memory leak, but tentatively dismissed it as [this](https://bugs.chromium.org/p/v8/issues/detail?id=10057) (also [here](https://github.com/nodejs/node/issues/28420)).

We are (obviously) no longer running the debug agent anywhere, and it isn't critical to our daily operations. Please let me know if you have any more questions or could use more information.

#### Environment details

  - OS: GKE container optimized
  - Node.js version: 10.17.0 - 10.18.1, 12.14.1, 13.6.0
  - `@google-cloud/debug-agent` version: 4.2.0 - 4.2.1

#### Steps to reproduce

1. `yarn add @google-cloud/debug-agent`
2. Add the following (with either value for `allowExpressions`):
   ```js
   // Add cloud debugging.
   import { start as startDebug } from "@google-cloud/debug-agent";
   startDebug({ serviceContext, allowExpressions: true });
   ```
3. Allow the application to run inside GKE with access to cloud debugging, serving a steady stream of requests over HTTP. The service graphed above was receiving between 1.5 and 3 requests per second.

**Comment by mctavish** [comment]
Sorry for the long silence on this issue.  I'm looking into this and #811 at the moment.

Thanks for all of the information; it's very much appreciated.

Do you know whether the cloud debugger was actively being used during the times that the performance degraded?  If there were no logpoints or snapshots being requested then it limits the potential codepaths that could be involved.

**Comment by mike-marcacci** [comment]
Hi @mctavish, I believe we saw this behavior irrespective of whether the debugger was actively being used for logpoints or snapshots.

**Comment by mctavish** [comment]
Thanks. That helps

**Comment by Louis-Ye** [comment]
We cannot reproduce the error performance wise, but according to the graph, it looks related to the memory leak in #811. Since we have came up the fix for the memory leak after v5.2.4, we would close this issue for now.  Please reopen it if the problem still exists.
